src/tool_writeout_json.c

```diff
@@ -0,0 +1,178 @@
+  {"num_connects", JSON_LONG, CURLINFO_NUM_CONNECTS},
+  {"time_redirect", JSON_DOUBLE, CURLINFO_REDIRECT_TIME},
+  {"num_redirects", JSON_LONG, CURLINFO_REDIRECT_COUNT},
+  {"ftp_entry_path", JSON_STRING, CURLINFO_FTP_ENTRY_PATH},
+  {"redirect_url", JSON_STRING, CURLINFO_REDIRECT_URL},
+  {"ssl_verify_result", JSON_LONG, CURLINFO_SSL_VERIFYRESULT},
+  {"proxy_ssl_verify_result", JSON_LONG, CURLINFO_PROXY_SSL_VERIFYRESULT},
+  {"filename_effective", JSON_FILENAME, CURLINFO_NONE},
+  {"remote_ip", JSON_STRING, CURLINFO_PRIMARY_IP},
+  {"remote_port", JSON_LONG, CURLINFO_PRIMARY_PORT},
+  {"local_ip", JSON_STRING, CURLINFO_LOCAL_IP},
+  {"local_port", JSON_LONG, CURLINFO_LOCAL_PORT},
+  {"http_version", JSON_VERSION, CURLINFO_HTTP_VERSION},
+  {"scheme", JSON_STRING, CURLINFO_SCHEME},
+  {NULL, JSON_NONE, CURLINFO_NONE}
+};
+
+static const char* http_version[] = {
+  "0",   /* CURL_HTTP_VERSION_NONE */
+  "1",   /* CURL_HTTP_VERSION_1_0 */
+  "1.1", /* CURL_HTTP_VERSION_1_1 */
+  "2"    /* CURL_HTTP_VERSION_2 */
+  "3"    /* CURL_HTTP_VERSION_3 */
+};
+
+static void json_escape(FILE* stream, const char* in)
+{
+  const char* i = in;
+  const char* in_end = in + strlen(in);
+
+  for (; i < in_end; i++) {
+    switch (*i) {
+    case '\\':
+      fputs("\\\\", stream);
+      break;
+    case '\"':
+      fputs("\\\"", stream);
+      break;
+    case '\b':
+      fputs("\\b", stream);
+      break;
+    case '\f':
+      fputs("\\f", stream);
+      break;
+    case '\n':
+      fputs("\\n", stream);
+      break;
+    case '\r':
+      fputs("\\r", stream);
+      break;
+    case '\t':
+      fputs("\\t", stream);
+      break;
+    default:
+      if (*i < 32) {
+        fprintf(stream, "u%04x", *i);
+      } else {
+        fputc(*i, stream);
+      }
+      break;
+    };
+  }
+}
+
+
+void ourWriteOutJSON(CURL *curl, struct OutStruct *outs, FILE *stream)
+{
+  char *stringp = NULL;
+  long longinfo;
+  double doubleinfo;
+  int i;
+
+  fputs("{", stream);
+  for (i = 0; keys[i].name != NULL; i++) {
+    const char* k = keys[i].name;
+    switch (keys[i].type) {
+    case JSON_STRING:
+      if((CURLE_OK == curl_easy_getinfo(curl, keys[i].cinfo, &stringp)) && stringp) {
+        fprintf(stream, "\"%s\":\"", k);
+        json_escape(stream, stringp);
+        fprintf(stream, "\",");
+      }
+      break;
+    case JSON_LONG:
+      longinfo = 0;
+      if(CURLE_OK == curl_easy_getinfo(curl, keys[i].cinfo, &longinfo)) {
+        fprintf(stream, "\"%s\":%ld,", k, longinfo);
+      }
+      break;
+    case JSON_DOUBLE:
+      if(CURLE_OK == curl_easy_getinfo(curl, keys[i].cinfo, &doubleinfo)) {
+        fprintf(stream, "\"%s\":%.6f,", k, doubleinfo);
+      }
+      break;
+    case JSON_FILENAME:
+      if(outs->filename) {
+        fprintf(stream, "\"%s\":\"", k);
+        json_escape(stream, outs->filename);
+        fprintf(stream, "\",");
+
+      } else {
+        fprintf(stream, "\"%s\":NULL,", k);
```

Comment:
Please make `NULL` lowercase according to https://stackoverflow.com/questions/21120999/representing-null-in-json